### DISCUSSION TITLE
Restart G6/Firefly with code 0. Error 4

### DISCUSSION THREAD

**Discussion by robertrub** [Navid200/xDrip#65] [Dexcom]
Hi @Navid200 As requested, here are the screen captures of the logs. 
5 min after the first error code 4, all was working ok.
![Screenshot_20210814-160516_xDrip+](https://user-images.githubusercontent.com/76626474/129448958-001db092-0328-4194-a1cb-88e7c143a11d.jpg)
![Screenshot_20210814-160541_xDrip+](https://user-images.githubusercontent.com/76626474/129448966-1dba817a-f904-4734-8a64-a7ffbca0ffef.jpg)
![Screenshot_20210814-161300_xDrip+](https://user-images.githubusercontent.com/76626474/129449100-ec7c8502-cd67-4de1-b1f0-580edf2bdf01.jpg)

Robert

**Comment by Navid200** [comment]
Thanks so much for posting this.

The two messages you see in your logs, at 15:25, one after the other come from the same if statement and only one should appear!
I am talking about "Calibration rejected ..." and "Calibration accepted ...".
I will look at this to see if I can understand why it happened and fix it if I can.
Thanks so much.

I used to have two options, original code and no code, in my restart guide.
Recently, I experienced a confused calibration error in no-code mode with the latest transmitter firmware.  I am not sure why this behavior as I had tested it before I put it in my guide.  However, I had tested it with a previous firmware.

Since I added the two options in my guide, I had got several comments, on facebook, about getting confused calibration in no-code mode.

Almost always, except for testing, I use the original code.

From those, I decided to remove the no-code mode from my guide:
https://github.com/Navid200/xDrip/wiki/Restart-G6-Sensor

You can continue to use the no-code mode if it works for you.  
I have to worry about supporting something and how best I can expect predictable results when I tell people what to do or not.
Sometimes, I have to remove something from my guide not because I am sure that it will fail.  But, because I cannot keep testing it and I cannot guarantee the outcome.

Please let me know if this makes no sense, or if you have any questions or suggestions.

**Comment by Navid200** [comment]
I noticed that you have posted a message in the issue I linked in the Gitter room.
Anyone is welcome to contribute to issues.
But, I hope that the issues platform will be treated really seriously and kept clean for the purpose of helping the developers improve/maintain the code.

The users need a platform to talk and rant and complain and cheer and share their experiences.  That's very important and required.  Most ideas come from looking at what users say.

Unfortunately, the only available platforms for users to share ideas have been the facbook group and the Gitter room.
https://www.facebook.com/groups/xDripG5
https://gitter.im/jamorham/xDrip-plus

So, many have been using the issues platform, on Github, in absence of a forum.
My hope is that people will use my discussions for that.

You are more than welcome to post in any issue.
But, I respectfully ask you to make sure what you post is relevant and have any detail required.  
For example, for your post in that issue to be useful, I ask you to please add the firmware number of the transmitter.  Without it, your post will just bring up more questions than answering any.
So, if you don't mind, please either add that info or just delete your post.

**Comment by Navid200** [comment]
If you like to keep that post in the issues platform, would you please also add the version of xDrip you are using to that post as well.  Please don't add more posts.  Please just edit the existing post and add the requested info.

**Comment by robertrub** [comment]
Hi

No problem, I deleted the message. I wasn't sure where you needed to have it ;)

Robert

----------------------------------------

14 août 2021 16:47:55 Navid ***@***.***>:

**Reply by Navid200** [reply]
It's perfect here.  
I am consulting with developers already to understand why both of those logs have been issued.  After we have an answer, I can figure out how to resolve it.
Thanks

**Comment by robertrub** [comment]
Hi

After thinking about this bug for a time, I might have an explanation.

I'm restarting a G6. After the 2h wait, I have the /calibration needed/ message then the window to enter the calibration.

I have started, at the same time, my Contour Next One and do a blood test. xDrip is set to use automatically the Contour BG results for calibration.

I do a blood test which is accepted by xDrip automatically AND the enter calibration page is gone. I think that xDrip is mightily well built as it used the blood calibration as the initial calibration :)

But, when I look at the system state, it shows */2/* waiting events and calibration 110 is the only line I can see. I don't know what the 2nd action is (maybe a feature request : to see the list of waiting actions there....)

Supposition : Is it possible that the calibration was treated 2 times, once for initial sensor start calibration and once a "normal" calibration. Maybe check the code to see what happens when initial calibration is needed and it comes via auto upload of BG from Contour Next One.

Best regards

Robert

----------------------------------------

14 août 2021 18:25:07 Navid ***@***.***>:

**Reply by Navid200** [reply]
I'm not sure if this is related to having calibration accepted and rejected both issued at the same time.
But, thanks for bringing this up.

Again, since you have not disclosed your transmitter's firmware or the first two characters of your transmitter's serial number, it's practically pointless to have a meaningful discussion about anything technical.  
There is a world of difference between an old and a new G6.

**Comment by Navid200** [comment]
Why do you have automatic calibration enabled for a G6?

**Comment by robertrub** [comment]
Hi

Sorry I missed the info telling that you needed the firmware :(

I don't have the FW with me and can't remember where it is shown... but below is the 8Pxxx number. Is ot sufficient?


 [cid:eu.faircode.email.18759]


----------------------------------------

15 août 2021 14:13:32 Navid ***@***.***>:

**Comment by Navid200** [comment]
Yes, that a Firefly.  Meaning, you can only run it in native mode.
In native mode, it is the transmitter that is responsible for calibration, not xDrip.
Automatic calibration would be a bad idea in native mode.

**Comment by robertrub** [comment]
Being a bit (!) perfectionist, I like my BG results as precise as possible and I calibrate my G6 (3-5 times over 10 days).

Without calibration (even during the initial 10 days), the results are usually higher than blood tests.

----------------------------------------

15 août 2021 15:11:25 Navid ***@***.***>:

**Comment by Navid200** [comment]
Calibrating and using automatic calibration are two different things.

A long time ago before there was G6, G5 could be used in non-native mode in xDrip and the result was good.

G6 brought factory calibration.  Using a G6 in non-native mode would give up factory calibration.  So, the pros of native mode overwhelms the pros of non-native mode for G6.    It didn't for G5.

Automatic calibration is a very interesting concept.
One of the most serious shortcomings of calibration is that you have no way of knowing what your blood glucose trend is when you calibrate.  All  you know is your interstitial fluid glucose trend.  Just because your interstitial fluid glucose trend is flat does not guarantee that your blood glucose trend would be flat too.  Remember that there is always a delay in interstitial fluid glucose relative to blood glucose.
So, a genius came up with the idea of automatic calibration:

The blood glucose reading is stored and is not submitted as calibration.  After 15 minutes, if the trend is still flat, it means that the blood glucose trend was also flat at the time of calibration.  Now, the calibration is applied but it is applied for 15 minutes ago when it was measured.
This can be done by xDrip because in non-native mode xDrip is responsible for calibration.  The transmitter has no idea what xDrip is doing with the raw values.

But, if you use automatic calibration in native mode, you are submitting the calibration to the transmitter with a time parameter.  We have no ideas what the transmitter does with the time parameter.  For all we know, the transmitter just drops the time parameter and uses the calibration at the time it has received it, which would be invalid because the reading is 15 minutes old.

I have to admit that I am now having a discussion with the developers to confirm this.
If I find out that I am mistaken about this, I will come back here and post here that I was wrong.

**Comment by robertrub** [comment]
If I understood correctly,  /auto calibration/ just takes the BG result from Countour Next One and sends it to the transmitter without my intervention. Does it do anything else ?

----------------------------------------

15 août 2021 15:56:50 Navid ***@***.***>:

**Reply by Navid200** [reply]
We may have posted exactly at the same time.  Please see my post above.

**Comment by robertrub** [comment]
Tnx a lot for the explanation. It explains why a "strange" message tells me that calibration will be applied in 15 min but it is applied as soon as the next connection to the transmitter.

I think that your supposition that the transmitter drops the time data is correct.

----------------------------------------

15 août 2021 16:17:00 Navid ***@***.***>:
